messaging/src/main/java/org/axonframework/commandhandling/gateway/AbstractRetryScheduler.java

```diff
@@ -0,0 +1,179 @@
+     */
+    protected abstract long computeRetryInterval(CommandMessage commandMessage,
+                                                 RuntimeException lastFailure,
+                                                 List<Class<? extends Throwable>[]> failures);
+
+    /**
+     * This is the entrypoint of the {@link RetryScheduler}. This default implementation checks if the last failure was
+     * transient, and if so reschedules a command dispatch.
+     *
+     * @param commandMessage The Command Message being dispatched
+     * @param lastFailure    The last failure recorded for this command
+     * @param failures       A condensed view of all known failures of this command. Each element in the array
+     *                       represents the cause of the element preceding it.
+     * @param dispatchTask   the task that performs the actual dispatching.
+     * @return {@code true} if rescheduling succeeded, {@code false} if otherwise.
+     */
+    @Override
+    public boolean scheduleRetry(CommandMessage commandMessage,
+                                 RuntimeException lastFailure,
+                                 List<Class<? extends Throwable>[]> failures,
+                                 Runnable dispatchTask) {
+        int failureCount = failures.size();
+        if (!isExplicitlyNonTransient(lastFailure) && failureCount <= maxRetryCount) {
+            if (logger.isInfoEnabled()) {
+                logger.info("Processing of Command [{}] resulted in an exception. Will retry {} more time(s)... "
+                                    + "Exception was {}, {}",
+                            commandMessage.getPayloadType().getSimpleName(),
+                            maxRetryCount - failureCount,
+                            lastFailure.getClass().getName(),
+                            lastFailure.getMessage()
+                );
+            }
+            return scheduleRetry(dispatchTask, computeRetryInterval(commandMessage, lastFailure, failures));
+        } else {
+            if (failureCount >= maxRetryCount && logger.isInfoEnabled()) {
+                logger.info("Processing of Command [{}] resulted in an exception {} times. Giving up permanently. ",
+                            commandMessage.getPayloadType().getSimpleName(), failureCount, lastFailure);
+            } else if (logger.isInfoEnabled()) {
+                logger.info("Processing of Command [{}] resulted in an exception and will not be retried. ",
+                            commandMessage.getPayloadType().getSimpleName(), lastFailure);
+            }
+            return false;
+        }
+    }
+
+    /**
+     * A builder class for {@link RetryScheduler}s. Values for the {@link Builder#retryExecutor} and
+     * {@link Builder#maxRetryCount} fields are both required, with the latter having a default value of
+     * {@link #DEFAULT_MAX_RETRIES}.
+     */
+    public static class Builder<B extends Builder> {
```

Comment:
Shouldn't the `Builder` be an `abstract class`, just like the class it's part of?

Reply:
Ok, added `abstract`. Unfortunately trying to add an abstract method `build()` will not work, as you don't have the concrete class available. Trying to add that as a generic parameter makes a bloody mess of the builder's signature, because you still have to keep the concrete builder as a parameter for the fluent methods. Trying to use only the concrete class built and using `RS.Builder` or something like that is beyond the capabilities of the type checker of the Java compiler, so that doesn't work. Where is 'Like Current' when you need it...